src/cockpit/channels/http.py

```diff
@@ -0,0 +1,153 @@
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import asyncio
+import http.client
+import logging
+import ssl
+import socket
+import threading
+
+from ..channel import Channel
+
+logger = logging.getLogger(__name__)
+
+
+class HttpChannel(Channel):
+    payload = 'http-stream2'
+
+    def create_connection(self):
+        opt_address = self.options.get('address') or 'localhost'
+        opt_port = self.options.get('port')
+        opt_unix = self.options.get('unix')
+        opt_tls = self.options.get('tls')
+        logger.debug('connecting to %s:%s; tls: %s', opt_address, opt_port or opt_unix, opt_tls)
+
+        if opt_tls is not None:
+            if 'authority' in opt_tls:
+                if 'data' in opt_tls['authority']:
+                    context = ssl.create_default_context(cadata=opt_tls['authority']['data'])
+                else:
+                    context = ssl.create_default_context(cafile=opt_tls['authority']['file'])
+            else:
+                context = ssl.create_default_context()
+
+            if 'validate' in opt_tls and not opt_tls['validate']:
+                context.check_hostname = False
+                context.verify_mode = ssl.VerifyMode.CERT_NONE
+
+            connection = http.client.HTTPSConnection(opt_address, opt_port, context=context)
+        else:
+            connection = http.client.HTTPConnection(opt_address, opt_port)
+
+        try:
+            if opt_unix:
+                # create the connection's socket so that it won't call .connect() internally (which only supports TCP)
+                connection.sock = socket.socket(socket.AF_UNIX)
+                connection.sock.connect(opt_unix)
+            else:
+                # explicitly call connect(), so that we can do proper error handling
+                connection.connect()
+        except (OSError, IOError) as e:
+            logger.error('Failed to open %s:%s: %s %s', opt_address, opt_port or opt_unix, type(e), e)
+            problem = 'unknown-hostkey' if isinstance(e, ssl.SSLCertVerificationError) else 'not-found'
+            self.close(problem=problem, message=str(e))
+            return None
+
+        return connection
+
+    def read_send_response(self, response):
+        '''Completely read the response and send it to the channel'''
+
+        while True:
+            # we want to stream data blocks as soon as they come in
+            block = response.read1(4096)
+            if not block:
+                logger.debug('reading response done')
+                # this returns immediately and does not read anything more, but updates the http.client's
+                # internal state machine to "response done"
+                block = response.read()
+                assert block == b''
+                break
+            logger.debug('read block of size %i', len(block))
+            self.loop.call_soon_threadsafe(self.send_data, block)
```

Comment:
So funny, enough, I *think* that calling `self.send_data()` is probably already currently threadsafe.  It basically ends up boiling down to a single write call on the underlying transport.

I don't want to guarantee that it would always be that way, but this shunting-over-to-the-main-thread on the main data path here is probably quite wasteful...

This needs a bit of a think.  I've already been contemplating moving away from Python's internal protocol/transport abstraction to a self-rolled thing that we have more control over.  This would remove the asyncstdio hackery and give us a chance to increase performance, and perhaps also make guarantees about threadsafety in places like this...

So for now: please keep it like it is :)

Reply:
I indeed started with .send_data(), with a big "XXX TODO" item that this didn't smell thread-safe. That's when I wrote the parallel torture unit test -- I fully expected that to break, but I wasn't actually able to. But that only raced against other http channel writes, and if that send_data() call is a single write that would explain why that worked. However, I feel very nervous to assume that *every* channel does "atomic" writes -- that smells like hard-to-debug problems later one.

So agreed that for now I feel better with keeping it as-is.